### PR TITLE
cisco_telemetry_mdt enhancement

### PR DESCRIPTION
**Pull request by dsx1123**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### Summary of Changes
1.  Fixed NXAPI to handle more complicated output
2.  DME add support for Events and addd support for Class based query
3.  Property transformation logicSummary of Changes

### CODE REVIEW

File: plugins/inputs/cisco_telemetry_mdt/README.md [hunk 1]
```diff
@@ -21,6 +21,9 @@ The TCP dialout transport is supported on IOS XR (32-bit and 64-bit) 6.1.x and l
  ## Address and port to host telemetry listener
  service_address = ":57000"
 
+ ## Grpc Maximum Message Size, default is 4MB, increase the size.
+ # max_msg_size = 20000000
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/README.md:25]: According to https://github.com/influxdata/telegraf/wiki/SampleConfig the commented-out option should have the default value.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package cisco_telemetry_mdt
 import (
 	"bytes"
 	"encoding/binary"
+        "encoding/json"
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:6]: Seems like formatting is mangled. Please use `make check` and `make fmt`. Also rebase your PR to current master to re-trigger the CI scripts.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 3]
```diff
@@ -16,8 +17,8 @@ import (
 	telemetry "github.com/cisco-ie/nx-telemetry-proto/telemetry_bis"
 	"github.com/golang/protobuf/proto"
 	"github.com/influxdata/telegraf"
+        internaltls "github.com/influxdata/telegraf/plugins/common/tls"
 	"github.com/influxdata/telegraf/metric"
-	internaltls "github.com/influxdata/telegraf/plugins/common/tls"
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:20]: Why?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 4]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
 	warned    map[string]struct{}
 	extraTags map[string]map[string]struct{}
+	nxpathMap map[string]map[string]string //per path map
+	nxpropMap map[string]int //Global property map
 	mutex     sync.Mutex
 	acc       telegraf.Accumulator
 	wg        sync.WaitGroup
 }
 
+type JsonStructure struct {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:65]: This is a really bad struct name especially due to the fact that it is exported! So please choose a name that describes what the structure contains and make it private.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 5]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
+	Name string `json:"Name"`
+	Prop []struct {
+		Key   string `json:"Key"`
+		Value string `json:"Value"`
+	} `json:"prop"`
+}
+
+//xform Field to string
+func xformValueString(field *telemetry.TelemetryField) interface{} {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:74]: Shouldn't the return value be of type `string`?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 6]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
+		str = strconv.FormatInt(val.Sint64Value, 10)
+		return str
+        }
+        return nil
+}
+
+//Xform value field.
+func (c *CiscoTelemetryMDT) nxosValueXform(field *telemetry.TelemetryField, value interface{}, path string) interface{} {
+     isNXOS := !strings.ContainsRune(path, ':')
+     if (isNXOS) {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:100]: Could you please fuse these two lines!

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 7]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
+        if xfrom, ok := c.nxpropMap[field.Name]; ok {
+           if xfrom == 1 {
+               switch field.ValueByType.(type) {
+               case *telemetry.TelemetryField_Uint64Value:
+                    return int64(value.(uint64))
+               default:
+                    return nil
+               }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:108]: No need for using a switch, a simple if should suffice. Furthermore, why don't you directly type-assert on value? It seems a bit fragile to blindly assume that `value` always follows the type specified in `field`...
Furthermore, you risk wrapping of values here by casting your `uin64`  to `int64`.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 8]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
+           } else if xfrom == 2 { //if type is uint64 convert it to string.
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:109]: Can we please `switch` on xfrom instead of doing the whole `else if` list!?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 9]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
+                switch val := field.ValueByType.(type) {
+                case *telemetry.TelemetryField_StringValue:
+                        if len(val.StringValue) > 0 {
+                           return val.StringValue
+                        }
+                case *telemetry.TelemetryField_Uint64Value:
+                        str := strconv.FormatUint(val.Uint64Value, 10)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:116]: You can directly `return` here.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 10]
```diff
@@ -50,13 +52,171 @@ type CiscoTelemetryMDT struct {
...
+                        return str
+                }
+                return nil
+           } else if xfrom == 3 {
+                //convert property to float from string.
+                switch val := field.ValueByType.(type) {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:122]: No need for `switch`...

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 11]
```diff
@@ -66,12 +226,96 @@ func (c *CiscoTelemetryMDT) Start(acc telegraf.Accumulator) error {
 		return err
 	}
 
+        //Create Nexus property transform list.
+        c.nxpropMap = make(map[string]int, 100 + len(c.Dmes))
+        c.nxpropMap["asn"] = 2 //check if it's uint64 type then convert to string.
+        c.nxpropMap["subscriptionId"] = 2 //check if it's uint64 type then convert to string.
+	c.nxpropMap["operState"] = 2 //check if it's uint64 type then convert to string.
+
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:234]: `c.nxpropMap` should really be a `map[string]func(field *telemetry.TelemetryField, value interface{}) interface{}` and contain function pointers to the conversions instead of this indirections using strings.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 12]
```diff
@@ -66,12 +226,96 @@ func (c *CiscoTelemetryMDT) Start(acc telegraf.Accumulator) error {
...
+{"Name": "sys/procsys","prop": [{"Key": "name","Value": "string"}, {"Key": "id","Value": "string"}, {"Key": "upTs","Value": "string"}, {"Key": "interval","Value": "string"}, {"Key": "memstatus","Value": "string"}]},
+{"Name": "sys/proc","prop": [{"Key": "processName","Value": "string"}, {"Key": "procArg","Value": "string"}]},
+{"Name": "sys/bfd/inst","prop": [{"Key": "descr","Value": "string"}, {"Key": "vrfName","Value": "string"}, {"Key": "name","Value": "string"}, {"Key": "name","Value": "string"}]},
+{"Name": "sys/lldp","prop": [{"Key": "sysDesc","Value": "string"}, {"Key": "portDesc","Value": "string"}, {"Key": "portIdV","Value": "string"}, {"Key": "chassisIdV","Value": "string"}, {"Key": "sysName","Value": "string"}, {"Key": "name","Value": "string"}, {"Key": "id","Value": "string"}]}]`)
+        var jdata []JsonStructure
+
+        c.dmes = make(map[string]string, len(c.Dmes))
+        c.nxpathMap = make(map[string]map[string]string, len(c.Dmes)+len(jdata))
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:278]: `len(jdata)` is zero here. I doubt that's what you wanted!?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 13]
```diff
@@ -66,12 +226,96 @@ func (c *CiscoTelemetryMDT) Start(acc telegraf.Accumulator) error {
...
+{"Name": "sys/intf","prop": [{"Key": "descr","Value": "string"}, {"Key": "name","Value": "string"}, {"Key": "lastStCause","Value": "string"}, {"Key": "description","Value": "string"}, {"Key": "unit","Value": "string"}, {"Key": "operFECMode","Value": "string"}, {"Key": "operBitset","Value": "string"}, {"Key": "mdix","Value": "string"}]},
+{"Name": "sys/procsys","prop": [{"Key": "name","Value": "string"}, {"Key": "id","Value": "string"}, {"Key": "upTs","Value": "string"}, {"Key": "interval","Value": "string"}, {"Key": "memstatus","Value": "string"}]},
+{"Name": "sys/proc","prop": [{"Key": "processName","Value": "string"}, {"Key": "procArg","Value": "string"}]},
+{"Name": "sys/bfd/inst","prop": [{"Key": "descr","Value": "string"}, {"Key": "vrfName","Value": "string"}, {"Key": "name","Value": "string"}, {"Key": "name","Value": "string"}]},
+{"Name": "sys/lldp","prop": [{"Key": "sysDesc","Value": "string"}, {"Key": "portDesc","Value": "string"}, {"Key": "portIdV","Value": "string"}, {"Key": "chassisIdV","Value": "string"}, {"Key": "sysName","Value": "string"}, {"Key": "name","Value": "string"}, {"Key": "id","Value": "string"}]}]`)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:274]: Can we please move this to its own file!? This clutters the file and makes reading difficult.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 14]
```diff
@@ -66,12 +226,96 @@ func (c *CiscoTelemetryMDT) Start(acc telegraf.Accumulator) error {
...
+        c.dmes = make(map[string]string, len(c.Dmes))
+        c.nxpathMap = make(map[string]map[string]string, len(c.Dmes)+len(jdata))
+        if err = json.Unmarshal(jsonData, &jdata); err == nil {
+            for _, jd := range jdata {
+                c.nxpathMap[jd.Name] = make(map[string]string, len(jd.Prop))
+                for _, prop := range jd.Prop {
+                    c.nxpathMap[jd.Name][prop.Key] = prop.Value
+                }
+            }
+        }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:286]: Why don't you directly define `nxpathMap` without the indirection of the JSON?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 15]
```diff
@@ -312,16 +558,17 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
 	for _, metric := range grouper.Metrics() {
 		c.acc.AddMetric(metric)
 	}
+        fmt.Println("Deepak: end", c.acc)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:561]: Please remove.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 16]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+            //For Every table fill the keys which are vrfName, address and masklen
+            if subfield.Name == "vrfName" {
+                tags["vrfName"] = decodeTag(subfield)
+            }
+            if subfield.Name == "address" {
+                tags["address"] = decodeTag(subfield)
+            }
+            if subfield.Name == "maskLen" {
+                tags["maskLen"] = decodeTag(subfield)
+            }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:654]:
```suggestion
            switch subfield.Name {
            case "vrfName",  "address", "maskLen" :
                tags["subfield.Name] = decodeTag(subfield)
            }
```

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 17]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+            if value := decodeValue(subfield); value != nil {
+                grouper.Add(measurement, tags, timestamp, subfield.Name, value)
+            }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:657]: Can you really rely on the order of `subfield.Name`s? Otherwise, depending on the order in `field.Fields`, you might emit metrics with different number of tags.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 18]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+            if subfield.Name == "nextHop" {
+                //For next hop table fill the keys in the tag - which is address and vrfname
+                for _, subf := range subfield.Fields {
+                    for _, ff := range subf.Fields {
+                        if ff.Name == "address" {
+                            tags["nextHop/address"] = decodeTag(ff)
+                        }
+                        if ff.Name == "vrfName" {
+                            tags["nextHop/vrfName"] = decodeTag(ff)
+                        }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:667]: Maybe also consolidate.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 19]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+        }
+        if field == nil {
+            return
+        }
+        if !isDme {
+            return
+        }
+
+        //TODO: Need to protect it from causing panic
+        if field.Fields[0].Fields[0].Fields[0].Name != "attributes" {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:700]: Right, this needs to be done. :-)

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 20]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+           //handle native data path rib
+           c.parseRib(grouper, field, prefix, path, tags, timestamp)
+           return
+        }
+        if field == nil {
+            return
+        }
+        if !isDme {
+            return
+        }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:697]: Fuse the two please.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 21]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+           return
+        }
+        nxAttributes = field.Fields[0].Fields[0].Fields[0].Fields[0]
+         
+        for _, subfield := range nxAttributes.Fields {
+                if subfield.Name == "rn" {
+                        rn = decodeTag(subfield)
+                } else if subfield.Name == "dn" {
+                        dn = true
+                        rn = decodeTag(subfield)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:710]: You might want to `break` here. Furthermore, I do not understand the `"rn"` branch. Again it depends on the order in `nxAttributes.Fields` which value `rn` will have. Is this intentional? Anyway `rn` will never be used if not entering the `dn` part right?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 22]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+                }
+            }
+        }
+}
+ 
+
+func (c *CiscoTelemetryMDT) parseClassAttributeField(grouper *metric.SeriesGrouper, field *telemetry.TelemetryField, prefix string, path string, tags map[string]string, timestamp time.Time) {
+        // DME structure: https://developer.cisco.com/site/nxapi-dme-model-reference-api/
+        rn := ""
+        dn := false
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:682]: Maybe move those down to the `for` loop filling the values.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 23]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+        if !dn { // Check for distinguished name being present
+                c.acc.AddError(fmt.Errorf("NX-OS decoding failed: missing dn field"))
+                return
+        }
+        tags["dn"] = rn
+
+        for _, subfield := range nxAttributes.Fields {
+                if subfield.Name != "dn" {
+                        c.parseContentField(grouper, subfield, "", path, tags, timestamp)
+                }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:723]: You could also collect the fields, tags etc. in the for loop above...

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 24]
```diff
@@ -391,9 +638,101 @@ func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemet
...
+        }
+
+        delete(tags, prefix)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:726]: Why is this needed?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 25]
```diff
@@ -430,11 +773,29 @@ func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, fie
 
 	var nxAttributes, nxChildren, nxRows *telemetry.TelemetryField
 	isNXOS := !strings.ContainsRune(path, ':') // IOS-XR and IOS-XE have a colon in their encoding path, NX-OS does not
+        var isEVENT bool
+        if isNXOS {
+           isEVENT = strings.Contains(path, "EVENT-LIST")
+        }
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:779]:
```suggestion
        isEVENT := isNXOS && strings.Contains(path, "EVENT-LIST")
```

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 26]
```diff
@@ -430,11 +773,29 @@ func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, fie
...
 	for _, subfield := range field.Fields {
 		if isNXOS && subfield.Name == "attributes" && len(subfield.Fields) > 0 {
 			nxAttributes = subfield.Fields[0]
 		} else if isNXOS && subfield.Name == "children" && len(subfield.Fields) > 0 {
-			nxChildren = subfield
+                        if !isEVENT {
+			    nxChildren = subfield
+                        } else {
+                            sub := subfield.Fields
+                            if sub[0].Fields[0].Name == "subscriptionId" && len(sub[0].Fields) >= 2  {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:788]: Might panic on subscriptions if empty.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 27]
```diff
@@ -430,11 +773,29 @@ func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, fie
...
+                                nxAttributes = sub[0].Fields[1].Fields[0].Fields[0].Fields[0].Fields[0].Fields[0]
+                            }
+                        }
+                        //if nxAttributes == NULL then class based query.
+                        if (nxAttributes == nil) {
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:793]: Remove parentheses.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 28]
```diff
@@ -450,9 +811,16 @@ func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, fie
 			for i, subfield := range row.Fields {
 				if i == 0 { // First subfield contains the index, promote it from value to tag
 					tags[prefix] = decodeTag(subfield)
+                                        //We can have subfield so recursively handle it.
+                                        if (len(row.Fields) == 1) {
+                                            tags["row_number"] = strconv.FormatInt(int64(i), 10)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:816]: This is overwritten seven lines down, so what is the point in setting this here?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 29]
```diff
@@ -450,9 +811,16 @@ func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, fie
...
+                                            c.parseContentField(grouper, subfield, "", path, tags, timestamp)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:817]: This is exactly the same function call as in the `else` branch. Why branching at all here?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 30]
```diff
@@ -513,24 +881,22 @@ const sampleConfig = `
  ## Telemetry transport can be "tcp" or "grpc".  TLS is only supported when
  ## using the grpc transport.
  transport = "grpc"
-
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:516]: Please keep the newlines...

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go [hunk 31]
```diff
@@ -238,6 +238,7 @@ func TestHandleNXAPI(t *testing.T) {
 	// error is expected since we are passing in dummy transport
 	require.Error(t, err)
 
+
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go:241]: No newline here...

**Review by srebhan — CHANGES_REQUESTED** [review]
Well this is actually quite hard to read. Especially the conversion handling in `nxosValueXform()` could greatly benefit from splitting the whole function into single conversion functions and directly map the field-names to those functions using `c.nxpropMap`.
Furthermore the construction of `c.nxpathMap` through the JSON-string is... unconventional. If there is no striking reason for this, please define the map directly and in a separate file to not impair readablility.
Last but not least, there is no test-case added as far as I can see. I regard this as must have considering the amount of code and complexity added.

**Comment by srebhan** [comment]
@dsx1123 I think something went wrong here. You end up with a lot of unrelated changes... Can you please rebase and revert the unrelated changes!?

**Comment by dsx1123** [comment]
It is rebased to upstream

**Comment by srebhan** [comment]
Check the files tab on the top! You are touching **228** files!!! I _think_ this is not what you intended!?

**Comment by dsx1123** [comment]
Not sure if I did something wrong but the the merge is rebase from upstream, none of the change is made from my side.

**Comment by srebhan** [comment]
There is a problem for sure. I cannot know which files you touched and which ones are from upstream. This renders any review impossible as nobody can go through 228 files. Maybe take a look at https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/syncing-a-fork and see if you can rebase according to this documentation...

**Comment by dsx1123** [comment]
got it, will revert the change and try to rebase again. thank for the info.

**Comment by dsx1123** [comment]
should have reverted change

**Comment by srebhan** [comment]
Still 228 files touched... Please, just read that link I posted and do a `git rebase` the correct way! You can easily check by looking at the files tab number at the top of this page. It should *not* show a number larger than say 20 or so...

**Comment by dsx1123** [comment]
it shows 4 only my side, not sure if I miss something. I will close this PR and create a new one with cleaner history.

**Comment by dsx1123** [comment]
one more thing, before I close this PR, one comment about indirection json file. Should I create map directly c.nxpathMap["show bgp ipv4"] = map[aspath:string] in separate function init_parser_db()?

**Comment by srebhan** [comment]
@dsx1123 Now it also shows only 4 touched files. Please keep this open in case you intend to still work on this!?
Regarding the indirection using the JSON file. I would like you to specify the resulting object structure directly. If I read between the lines ;-) the JSON file specifies an internal definition and you want to "generate" the object structure using this JSON-definition. If so, please write a small generator to *offline* generate the GO-code for that structure.
In case the JSON is written "manually", you might also directly "manually" generate the GO structure code.

The problems regarding the JSON way are, either you put the JSON as "text" into the code which bloats the code **or** we somehow need to ship that JSON file to the customer. For the latter there is not infrastructure and the former is just ugly. ;-)

**Comment by dsx1123** [comment]
@srebhan I have to close this PR and open a new one, since original forked repo has been removed. I have already forked a new one, sorry for messing up the "rebase". We should have fixed all the comment but only this one is left, will put new PR # here once it is opened alone with the fix of json file.

**Comment by srebhan** [comment]
Ok.

**Comment by dsx1123** [comment]
@srebhan  created new PR #8661, closing this one.
